File: LeetCode/189RotateArray/RotateArray.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        // Reverse -> Time: O(n), Space:O(1)
        k = k%nums.size();
        int temp, l = 0, r = nums.size()-1;
        while (l < r) {
            temp = nums[l];
            nums[l] = nums[r];
            nums[r] = temp;
            l++;
            r--;
        }
        l = 0;
        r = k-1;
        while (l < r) {
            temp = nums[l];
            nums[l] = nums[r];
            nums[r] = temp;
            l++;
            r--;
        }
        l = k;
        r = nums.size()-1;
        while (l < r) {
            temp = nums[l];
            nums[l] = nums[r];
            nums[r] = temp;
            l++;
            r--;
        }

        // // Temp array -> Time: O(n), Space:O(n)
        // int n = nums.size();
        // vector<int> new_nums(n);
        // for (int i = 0; i < n; i++) {
        //     new_nums[(i+k)%n] = nums[i];
        // }
        // for (int i = 0; i < n; i++) {
        //     nums[i] = new_nums[i];
        // }
    }
};

File: LeetCode/189RotateArray/RotateArray.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        # Reverse -> Time: O(n), Space:O(1)
        k = k%len(nums)
        l, r = 0, len(nums)-1
        while l<r:
            nums[l], nums[r] = nums[r], nums[l]
            l += 1
            r -= 1
        l, r = 0, k-1
        while l<r:
            nums[l], nums[r] = nums[r], nums[l]
            l += 1
            r -= 1
        l, r = k, len(nums)-1
        while l<r:
            nums[l], nums[r] = nums[r], nums[l]
            l += 1
            r -= 1
        
        # # Temp array -> Time: O(n), Space:O(n)
        # n = len(nums)
        # new_nums = [0 for i in range(n)]
        # for i in range(n):
        #     new_nums[(i+k)%n] = nums[i]
        # for i in range(n):
        #     nums[i] = new_nums[i]
